### PR TITLE
GET admin/user route

### PR DESCRIPTION
**Pull request by voidbuf**
get users based on whether the current user is an admin or not

### CODE REVIEW

**Comment by voidbuf** [comment]
[Admin route for getting the all users](https://trello.com/c/tE5OVBCJ/19-admin-route-for-getting-the-all-users)

File: controllers/userController.js [hunk 1]
```diff
@@ -16,6 +17,51 @@ exports.getAllUsers = async (req, res) => {
   }
 }
 
+exports.getUsers = async (req, res) => {
+
+  //get the current user id from the token
+  const token = getEvenToken(req)
+  const decoded = decodeTkn(token, req.config.JWT_KEY)
+  const id = decoded[constants.idClaim]
```
Comment (procateodor) [inline controllers/userController.js]: this parts is already in `requireAuth` middleware. So we can remove lines 22-25

File: controllers/userController.js [hunk 2]
```diff
@@ -16,6 +17,51 @@ exports.getAllUsers = async (req, res) => {
...
+
+  //check if the current user is an admin
+  let isAdmin = req.user.admin
+  
+  if(!isAdmin) {
+    //get the data of the current user
+    try {
+      const userData = await req.db.User.findById(id)
```
Comment (procateodor) [inline controllers/userController.js]: this should be `await req.db.User.findById(ObjectId(req.user._id))`

File: controllers/userController.js [hunk 3]
```diff
@@ -16,6 +17,51 @@ exports.getAllUsers = async (req, res) => {
...
+
+      return res.status(HttpStatus.OK).json({
+        success: true,
+        userData
```
Comment (procateodor) [inline controllers/userController.js]: we should return `user` instead of `userData`

File: controllers/userController.js [hunk 4]
```diff
@@ -16,6 +17,51 @@ exports.getAllUsers = async (req, res) => {
...
+      })
+    } catch (error) {
+      req.log.error(`Unable to get current user -> ${req.url} -> ${error}`)
+      return res.status(HttpStatus.INTERNAL_SERVER_ERROR).json({
+        success: false
+      })
+    }
+  }
+
+  if(isAdmin) {
```
Comment (procateodor) [inline controllers/userController.js]: no need of this if

File: controllers/userController.js [hunk 5]
```diff
@@ -16,6 +17,51 @@ exports.getAllUsers = async (req, res) => {
...
+    //get the data of all users based on query parameters
+    try {
+      const users = await req.db.User.find(req.query)
```
Comment (procateodor) [inline controllers/userController.js]: please make sure this is working even tho I pass no data in query params.

**Review by procateodor — CHANGES_REQUESTED** [review]
Some minor changes, and please pull master into your branch and resolve conflicts if so.

File: controllers/userController.js (hunk 3 above)
Comment (voidbuf) [reply controllers/userController.js]: If we return req.user, should we still do:
`await req.db.User.findById(ObjectId(req.user._id))`
or is it ok to just return req.user if the user is not an admin?
`if(!isAdmin)
return res.status(HttpStatus.OK).json({
        success: true,
        req.user
      })`
But when the user updates his/her data does a middleware update the token?

File: controllers/userController.js (hunk 5 above)
Comment (voidbuf) [reply controllers/userController.js]: Made sure it has the expected behavior. It returns all the users

File: controllers/userController.js (hunk 2 above)
Comment (voidbuf) [reply controllers/userController.js]: the req.user object doesn't seem to have the _id field, but I think I should use req.user['user:id']

**Review by procateodor — APPROVED** [review]

**Pull request merged**